synapse/handlers/room_member.py

```diff
@@ -218,6 +220,52 @@ def _local_membership_update(
                 newly_joined = prev_member_event.membership != Membership.JOIN
             if newly_joined:
                 yield self._user_joined_room(target, room_id)
+
+            # Copy over direct message status and room tags if this is a join
+            # on an upgraded room
+
+            # Check if this is an upgraded room
+            state_ids = yield self.store.get_current_state_ids(room_id)
+            create_id = state_ids.get((EventTypes.Create, ""))
+            if not create_id:
+                return
+            create_event = yield self.store.get_event(create_id)
+
+            if "predecessor" in create_event["content"]:
+                old_room_id = create_event["content"]["predecessor"]["room_id"]
+
+                # Retrieve room account data for predecessor room
+                user_account_data = yield self.store.get_account_data_for_user(
+                    user_id,
+                )
+                room_tags = yield self.store.get_tags_for_room(
+                    user_id, old_room_id,
+                )
+
+                # Copy direct message state if applicable
+                if user_account_data and "m.direct" in user_account_data[0]:
```

Comment:
suggest you do `direct_rooms = user_account_data.get("m.direct", {})` and you can drop these conditions.

HOWEVER it may be worth sanity-checking that nobody has decided to make `m.direct` a list or something else other than a dict, so `if isinstance(direct_rooms, dict)`.